File: Resources/source/EditScriptViewController.lua


local EditScriptViewController = {}


EditScriptViewController.mt = { __call = function(self, ...)
													return self.new (...)
											end
							  }

setmetatable (EditScriptViewController,EditScriptViewController.mt)

function EditScriptViewController.new (onDone, filename,...)
		
		local title = "Edit Script";
		if (String.IsNullOrEmpty (filename)) then
			title = "New Script";
		end

		EditScriptViewController.m = {}; -- create a table members to store Lua fields.
		EditScriptViewController.m.doneAction = onDone;
		
		luanet.make_object (EditScriptViewController, 'MonoTouch.Dialog.DialogViewController');

		EditScriptViewController.m.nameEntry = EntryElement ("Name", "Enter script name", filename);
		EditScriptViewController.Root = RootElement(title);
		local section = Section("script");
		section:Add(EditScriptViewController.m.nameEntry);
		EditScriptViewController.Root:Add(section);

		return EditScriptViewController;
end

function EditScriptViewController:ViewDidLoad ()
	self.base:ViewDidLoad();

	self.NavigationItem:SetLeftBarButtonItem(UIBarButtonItem (UIBarButtonSystemItem.Cancel, function () self:OnCancel() end), false);
	self.NavigationItem.RightBarButtonItem = UIBarButtonItem (UIBarButtonSystemItem.Done, function () self:OnDone() end);
	self.m.nameEntry:BecomeFirstResponder (false);
end

function EditScriptViewController:OnCancel ()
	self:DismissViewController(true, nil);
end

function EditScriptViewController:OnDone ()
	if (self.m.doneAction ~= nil) then
		self.m.doneAction (self.m.nameEntry.Value, function () self:DismissViewController (true, nil) end);
	end
end

return EditScriptViewController;

File: NLuaBox/ViewControllers/EditScriptViewController.cs
using Foundation;
using System;
using UIKit;

namespace NLuaBox
{
    public partial class EditScriptViewController : UITableViewController
    {
        public Action<EditScriptViewController, string> DoneAction { get; set; }

        string initialScriptName = string.Empty;

        public EditScriptViewController(IntPtr handle) : base(handle)
        {
        }

        public static EditScriptViewController Create(Action<EditScriptViewController, string> onDone)
        {
            return Create(onDone, string.Empty);
        }

        public static EditScriptViewController Create(Action<EditScriptViewController, string> onDone, string fileName)
        {
            UIStoryboard mainStoryboard = UIStoryboard.FromName("Main", null);
            EditScriptViewController viewController = (EditScriptViewController)mainStoryboard.InstantiateViewController("EditScriptViewController");

            if (string.IsNullOrEmpty(fileName))
                viewController.Title = NSBundle.MainBundle.GetLocalizedString("New_Script", "New Script");
            else
                viewController.Title = NSBundle.MainBundle.GetLocalizedString("Edit_Script", "Edit Script");

            viewController.DoneAction = onDone;
            viewController.initialScriptName = fileName;
            return viewController;
        }

        void ScriptName_ValueChanged(object sender, EventArgs e)
        {
            NavigationItem.RightBarButtonItem.Enabled = !string.IsNullOrEmpty(scriptName.Text) && scriptName.Text != initialScriptName;
        }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();

            NavigationItem.LeftBarButtonItem = new UIBarButtonItem(UIBarButtonSystemItem.Cancel, OnCancel);
            NavigationItem.RightBarButtonItem = new UIBarButtonItem(UIBarButtonSystemItem.Done, OnDone);
            NavigationItem.RightBarButtonItem.Enabled = false;
            scriptName.EditingChanged += ScriptName_ValueChanged;
            scriptName.WeakDelegate = this;
            scriptName.Text = initialScriptName;
            scriptName.BecomeFirstResponder();
        }

        void OnCancel(object sender, EventArgs e)
        {
            Dismiss();
        }

        [Export("textFieldShouldReturn:")]
        void TextFieldShouldReturn(UITextField field)
        {
            if (string.IsNullOrEmpty(scriptName.Text))
                return;

            OnDone(this, EventArgs.Empty);
        }

        void OnDone(object sender, EventArgs e)
        {
            if (DoneAction == null)
                return;

            DoneAction.Invoke(this, scriptName.Text);
        }

        public void Dismiss()
        {
            DismissViewController(true, null);
        }
    }
}